### ISSUE TITLE
Significantly improved MPI performance with revised partitioning

### ISSUE THREAD

**Issue by JamiePringle** [OceanParcels/parcels#1300]
I have a large problem to run, with 781,197 release points every half day for a year, each particle running for 60 days, so a total of 93,743,640 particles running simultaneously and 664,798,647 total particles over an entire hemisphere. (interested why? see https://github.com/JamiePringle/EZfate). These runs take a while, and in general are mostly CPU limited, not IO limited. 

I have found that I can improve the run time by 30% by altering the initial partitioning of the particles.  The stock code uses a means clustering algorithm to partition the particles into different MPI ranks. The MPI rank with the most particles has 2.6 times the average number of particles/rank, and the smallest is 0.22 times the mean. The latter MPI rank will wait doing nothing while the former finishes its job. (The smallest rank is spatially clustered around isolated island chains -- for example Hawaii).

By re-writing the initial partition to use a constrained k-means scheme that enforces that clusters be roughly equal size, I can get 30 to 40% faster run times (the speed-up is not 2.6 because in the initial unbalanced case, the largest ranks finish running on a lightly loaded machine). This is significant when runs can take 4 days.  My solution is not a panacea -- see the comments below. 

But I suggest for now that we change the code so that users can specify the particle partitioning. Maybe, in the future, we can change the default partitioning scheme. 

To get the improved performance with equal-sized jobs for each MPI rank, I use https://joshlk.github.io/k-means-constrained/ and replace the existing partitioning code in `collectionsoa.py` 
```
                                coords = np.vstack((lon, lat)).transpose()
                                kmeans = KMeans(n_clusters=mpi_size, random_state=0).fit(coords)
                                self._pu_indicators = kmeans.labels_
```
Unfortunately, the constrained k-means is much slower (more on that below), so I compute the clusters on a culled subset of the initial points, and then use the clustering to classify all the starting points. The resulting code becomes
```
                                #This code does a constrained k-means which ensures that each
                                #cluster has a the same number of particles to within a multiplicative factor "slop".
                                #Because the constrained k-means can be slow, the clustering is done on
                                #a data set that has been reduced by a factor of nCull. This should be determined
                                #automagically in a final version.
                                nCull=20
                                print('Starting constrained k-means for',len(lon),'particles and nCull',nCull,flush=True)
                                coords = np.vstack((lon, lat)).transpose()
                                coordsTrain=coords[::nCull,:]
                                slop=0.1
                                maxSizeCluster=int((1.0+slop)*(coordsTrain.shape[0]/mpi_size))
                                minSizeCluster=int((1.0-slop)*(coordsTrain.shape[0]/mpi_size))
                                kmeans = KMeansConstrained(n_clusters=mpi_size, size_min=minSizeCluster, size_max=maxSizeCluster,
                                                random_state=0,n_jobs=-4).fit(coordsTrain)
                                #now predict where all particles go from the kmeans calculated on partial data set
                                self._pu_indicators = kmeans.predict(coords,minSizeCluster*nCull,maxSizeCluster*nCull)
                                print('   done with constrained k-means',flush=True)
```
The resulting sets of starting points for the MPI ranks are all the same size to within 10% (as specified in variable `slop`).  This code only takes 25s to run, which is small compared to the total run time of the code (the original K-means finishes in about a second). 

So should we just drop in this revised code to the main branch?  I think that might be premature for two reasons:

- the KMeans code from sci-kit learn is from a well maintained project that will be around and stable for a long time; k-means-constrained is run by a single person (Josh Levy-Kramer) and while currently maintained is not part of a larger project. I have not talked to the author. But perhaps it is unwise to tie the project to a single-persons project?
- The constrained Kmeans code is much, much slower than the straight Kmeans code. Hence the nCull parameter above. One could imagine where an arbitrary choice of nCull could go wrong. 

As a suggestion, for now, perhaps it would be better to allow the user to specify a function `initialPartition(lon,lat,mpi_size)` that returns a vector of integers of size lon.shape in range(mpi_size) which specifies the initial partitioning of the particles, so that all this weird and situation dependent partitioning can, at least for now, be outside of the main code base?  If this function were passed in, it could be implemented within `collectionsoa.py` as
```
                                self._pu_indicators = initialPartition(lon,lat,mpi_size)
```
The default function would be based on the original primitive-but-very-robust Kmeans method.

If this seems reasonable to people, I could take a hack at it. 
Jamie

**Comment by erikvansebille** [comment]
Thanks for this extensive comment @JamiePringle, and I fully see your point. I actually like your proposed solution to allow users to override the default clustering method; so feel free to create a PR that implements this. Should not be too difficult, right?

**Comment by erikvansebille** [comment]
Note, by the way, that there is also a relevant open (stalled) PR at https://github.com/OceanParcels/parcels/pull/1231. Perhaps some of the ideas from there could also be incorporated?

**Comment by JamiePringle** [comment]
I think it would be straightforward to implement this in a way that does not break any existing code; I will have time late next week to take a hack at it. Jamie

**Comment by erikvansebille** [comment]
This is now now fixed in #1414 and #1424
